EIP-5173: NFT Future Rewards (nFR) Standard (#5173)
ethereum/EIPs: EIPS/eip-5173.md

---
eip: 5173
title: NFT Future Rewards (nFR) Standard
description: A multigenerational reward mechanism that rewards‌ all ‌owners of non-fungible tokens (NFT).

> author: Yale ReiSoleil <yale@iob.fi>, dRadiant (@dRadiant), D Wang, PhD <david@iob.fi>

Reviewer: That's awesome! Thank you for your time and suggestions. I'll incorporate your changes.

discussions-to: https://ethereum-magicians.org/t/non-fungible-future-rewards-token-standard/9203
type: Standards Track
category: ERC
status: Draft
created: 2022-05-08
requires: 165, 721
---

## Abstract

In this EIP, we propose the implementation of a Future Rewards (FR) extension which will enable owners of [EIP-721](./eip-721.md) tokens (NFTs) to participate in future price increases after they sell their tokens.

Owners of NFTs can expect to make money in two ways:

1. An increase in price during their holding period;
2. They receive Future Rewards (FRs) in the form of proceeds of the realized profits from the subsequent new owners after they have sold it. 

In the event the seller is not the first owner, the original minter, the profits gained when selling an NFT will be shared with the previous owners. The same person will receive the same FR distributions under the nFR system as the next generation of new owners after them. 

## Motivation

> Are you interested in finding something that may prove valuable in the future and acquiring it early? Excellent! In reality, you often find yourself in a predicament where it does not matter whether you are a paper hand trader or a diamond hand hodler, the price keeps rising. 

Reviewer: Thank you for your review and instructions.

NFTs, in contrast to physical art and collectibles in the physical world, are not currently reflecting the contributions of their owners to their value. Since the ERC721 token can be tracked individually, and may be modified to record every change in price of any specific NFT token, there is no reason that a Future Rewards program of this type should not be established.

This EIP establishes a standard interface for a profit sharing structure in all stages of the token's ownership history desired by all market participants.

Additionally, as we will explain later, it discourages any "under-the-table" deals that may circumvent the rules set forth by artists and marketplaces.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

We are implementing this extension using the  ERC721 standard.

> ERC721-compliant contracts MAY implement this EIP for rewards to provide a standard method of rewarding future buyers and previous owners with realized profits in the future.

Reviewer 2: ```suggestion
[EIP-721](./eip-721.md)-compliant contracts MAY implement this EIP for rewards to provide a standard method of rewarding future buyers and previous owners with realized profits in the future.
```

This system does not discriminate against any buyer because each buyer will go through the same distribution curve.

#### Straight line arithmetic sequence FR payout

![Figure 5: Arithmetic sequence distribution](../assets/eip-5173/Arithmetic_Sequence_FR_Payout_Distribution.png?raw=true)

*Figure 5: Arithmetic sequence distribution*

The profit is distributed according to the arithmetic sequence, which is 1, 2, 3, ... and so on. The first owner will receive 1 portion, the second owner will receive 2 portions, the third owner will receive 3 portions, etc. 

## Backwards Compatibility

This proposal is fully compatible with current [EIP-721](./eip-721.md) standards and [EIP-2981](./eip-2981.md). It can also be easily adapted to work with [EIP-1155](./eip-1155.md).

## Test Cases

> Following is a contract that detects which interfaces other contracts implement, from unTrading by @dRadiant at iob.fi DAO.

Reviewer: Thank you for your review and instructions.

The repository with the reference implementation contains all the tests.

[Here is an illustration of a test case](../assets/eip-5173/animate-1920x1080-1750-frames.gif?raw=true). 

## Reference Implementations

This implementation uses OpenZeppelin contracts and the PRB Math library created by Paul R Berg for fixed-point arithmetic. It demonstrates the interface for the nFR standard, an nFR standard-compliant extension, and an [EIP-721](./eip-721.md) implementation using the extension.

> GitHub Link: [https://github.com/dRadiant/nfr-reference-implementation](https://github.com/unTrading/eip-5173-reference-implementation)

Reviewer: Thank you for your review and instructions.

> GitHub Link: [https://github.com/dRadiant/nfr-reference-implementation](https://github.com/unTrading/eip-5173-reference-implementation)

Reviewer: Updated.

- n is the actual number of owners eligible and participating in the future rewards sharing. To calculate n, we have n = min(m, w), where m is the current number of owners for a token, and w is the window size of the profit distribution sliding window algorithm

#### Converting into Code

```solidity

pragma solidity ^0.8.0;
//...

/* Assumes usage of a Fixed Point Arithmetic library (prb-math) for both int256 and uint256, and OpenZeppelin Math utils for Math.min. */
function _calculateFR(uint256 P, uint256 r, uint256 g, uint256 m, uint256 w) pure internal virtual returns(uint256[] memory) {
        uint256 n = Math.min(m, w);
        uint256[] memory FR = new uint256[](n);

        for (uint256 i = 1; i < n + 1; i++) {
            uint256 pi = 0;

            if (successiveRatio != 1e18) {
                int256 v1 = 1e18 - int256(g).powu(n);
                int256 v2 = int256(g).powu(i - 1);
                int256 v3 = int256(P).mul(int256(r));
                int256 v4 = v3.mul(1e18 - int256(g));
                pi = uint256(v4 * v2 / v1);
            } else {
                pi = P.mul(r).div(n);
            }

            FR[i - 1] = pi;
        }

        return FR;
}

```

> The complete implementation code can be found [here](https://github.com/unTrading/eip-5173-reference-implementation).

Reviewer: Updated.

## Security Considerations

### Payment Attacks

> As this EIP introduces royalty and realized profit rewards collection, distribution, and payouts to the [EIP-721](./eip-721.md) standard, the attack vectors increase. As discussed in EIP-4910 and other mitigations to phishing attacks by Andreas Freund, we recommend reentrancy protection for all payment functions to reduce the most significant attack vectors for payments and payouts.

Reviewer 3: Probably shouldn't be referencing EIPs that haven't been merged into the repo yet.

Reviewer: Thank you for your review and instructions.

Reviewer: Updated.

#### Wash trading by users
Using a different wallet address, an attacker can "sell" the NFT to themselves at a loss. It is possible to repeat this process n times in order to maximize their share of the subsequent FR distributions (*Figure 6*). A wallet ranking score can partially alleviate this problem. It is evident that a brand new wallet is a red flag, and the marketplace may withhold FR distribution from it if it has a short transaction history (i.e. fewer than a certain number of transactions).

We do not want a large portion of future rewards to go to a small number of wash traders. Making such practices less profitable is one way to discourage wash trading and award hoarding. It can be partially mitigated, for example, by implementing a wallet-score and holding period-based incentive system. The rewards for both parties are reduced if a new wallet is used or if a holding period is less than a certain period. 

![Figure 6: Same owner using different wallets](../assets/eip-5173/5%20losing.jpeg?raw=true)

*Figure 6: Same owner using different wallets*

#### Wash trading by the marketplace operator

> However, the biggest offender appears to be the marketplace, which engages heavily in wash trading, or simply does not care about it, according to Decrypt. I have personally experienced this phenomenon. A senior executive of a top-5 cryptocurrency exchange boasted during a mid-night drinking session in 2018, that they had "brushed" (wash-traded) certain newly listed tokens, which they called "marketmaking." The exchange is still ranked among the top five crypto exchanges today.

Reviewer 3: Please don't use first person in EIPs.

Reviewer: Thank you for your review and instructions.

Reviewer: Updated. Thanks!

---
eip: 5173

> title: NFT Future Rewards (nFR) Standard

Reviewer 4: We try to discourage the word "standard" in titles and descriptions, because it tends to end up in every ERC.

```suggestion
title: NFT Future Rewards (nFR)
```

You'll probably also want to take a peek at https://github.com/ethereum/EIPs/runs/7393609255?check_suite_focus=true to see the other problems the linter is reporting.

description: A multigenerational reward mechanism that rewards‌ all ‌owners of non-fungible tokens (NFT).

> author: Yale ReiSoleil <@longnshort>, dRadiant (@dRadiant), D Wang, PhD <david@iob.fi>

Reviewer 5: ```suggestion
author: Yale ReiSoleil (@longnshort), dRadiant (@dRadiant), D Wang, PhD <david@iob.fi>
```

discussions-to: https://ethereum-magicians.org/t/non-fungible-future-rewards-token-standard/9203
type: Standards Track
category: ERC
status: Draft
created: 2022-05-08
requires: 165, 721
---

## Abstract

In this EIP, we propose the implementation of a Future Rewards (FR) extension which will enable owners of [EIP-721](./eip-721.md) tokens (NFTs) to participate in future price increases after they sell their tokens.

Through the implementation of EIP-5173, the creators, buyers and sellers create a giving circle in trading practice. A giving circle is formed when all participants work in a framework to build greater wealth through each other's success. One does not expect the same amount of return from the same person when they give someone a portion of their profits. There is no quid pro quo. Rather, they are confident that someone else in the same circle will give them the same benefits, with a smaller or greater monetary value, from other participants in the same circle later on.

> Owners of the EIP-5173 compliant tokens can benefit in two ways from such a gift economic framework:

Reviewer 4: Prefer "this proposal" instead of referring to it with its own number.

As in trading, most traders lose money, but the proposed Future Rewards framework is designed to help average traders do better.

Additionally, as we will explain later, it discourages any "under-the-table" deals that may circumvent the rules set forth by artists and marketplaces.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

The following is an extension of the [EIP-721](./eip-721.md) standard.

[EIP-721](./eip-721.md)-compliant contracts MAY implement this EIP for rewards to provide a standard method of rewarding future buyers and previous owners with realized profits in the future.

Implementers of this standard MUST have all of the following functions:

```solidity

pragma solidity ^0.8.0;

import "@openzeppelin/contracts/utils/introspection/IERC165.sol";

/*
 *
 * @dev Interface for the Future Rewards Token Standard.
 *
 * A standardized way to receive future rewards for non-fungible tokens (NFTs.)
 *
 */

> interface InFR is IERC165 {

Reviewer 4: It's customary to include the interface id explicitly somewhere in the EIP.

    event FRClaimed(address indexed account, uint256 indexed amount);

    event FRDistributed(uint256 indexed tokenId, uint256 indexed soldPrice, uint256 indexed allocatedFR);

    function list(uint256 tokenId, uint256 salePrice) external;

    function unlist(uint256 tokenId) external;

    function buy(uint256 tokenId) payable external;

    function releaseFR(address payable account) external;

    function retrieveFRInfo(uint256 tokenId) external returns(uint8, uint256, uint256, uint256, uint256, address[] memory);

    function retrieveAllottedFR(address account) external returns(uint256);

    function retrieveListInfo(uint256 tokenId) external returns(uint256, address, bool);
    
}

interface IERC165 {
    /// @notice Query if a contract implements an interface
    /// @param interfaceID The interface identifier, as specified in ERC-165
    /// @dev Interface identification is specified in ERC-165. This function
    ///  uses less than 30,000 gas.
    /// @return `true` if the contract implements `interfaceID` and
    ///  `interfaceID` is not 0xffffffff, `false` otherwise
    function supportsInterface(bytes4 interfaceID) external view returns (bool);

> }

Reviewer 4: You can omit this, since you've marked 165 as a required EIP.

The future Rewards payments SHOULD utilize a pull-payment model, similar to that demonstrated by OpenZeppelin with their PaymentSplitter contract. The event  FRClaimed would be triggered after a successful claim has been made. 

```solidity

function releaseFR(address payable account) public virtual override {
        //...
}

```

### Owner Generation Shifting

The `_shiftGenerations` function MUST be called regardless of whether the sale was profitable or not. As a result, it will be called in the `_transfer` [EIP-721](./eip-721.md) override function and the FR `transferFrom` function. The function SHALL remove the oldest account from the corresponding `_addressesInFR` array. This calculation will take into account the current length of the array versus the total number of generations for a given token ID.

## Rationale

### Is This Just a Ponzi Scheme? 

No, it is not. Ponzi schemes promise profits that are impossible to keep. 

As opposed to fixed-yield schemes, our proposal only distributes future profits when those profits are achieved rather than guaranteeing them. Should later holders fail to make a profit, future return shares will not be distributed. 

> The early participants in price discovery will receive a share of profits as part of the FR implementation only and if a later owner has accumulated profits during their holdings of the token.

Reviewer 4: This is more Motivation than Rationale.

The Motivation section describes the reason the EIP as a whole exists, while the Rationale section documents the reasoning behind specific technical choices within the EIP.

This system does not discriminate against any buyer because each buyer will go through the same distribution curve.

#### Straight line arithmetic sequence FR payout

![Figure 5: Arithmetic sequence distribution](../assets/eip-5173/Arithmetic_Sequence_FR_Payout_Distribution.png?raw=true)

*Figure 5: Arithmetic sequence distribution*

The profit is distributed according to the arithmetic sequence, which is 1, 2, 3, ... and so on. The first owner will receive 1 portion, the second owner will receive 2 portions, the third owner will receive 3 portions, etc. 

## Backwards Compatibility

This proposal is fully compatible with current [EIP-721](./eip-721.md) standards and [EIP-2981](./eip-2981.md). It can also be easily adapted to work with [EIP-1155](./eip-1155.md).

## Test Cases

> [This folder](../assets/eip-5173/Implementation) contains the reference implementation for EIP-5173.

Reviewer 4: Unfortunately our markdown renderer can't handle links to entire folders, so you'll have to pick a single file to link to.
